python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+            or `DimensionGraph`.
+        """
+        names = set(self.names)
+        universe = None
+        for other in others:
+            names &= toNameSet(other).names
+            universe = getattr(other, "universe", universe)
+        if universe is not None:
+            return DimensionSet(universe, names)
+        else:
+            return DimensionNameSet(names)
+
+    def symmetric_difference(self, other):
+        """Return a new set containing all elements that are in either ``self``
+        or other, but not both.
+
+        Parameters
+        ----------
+        other : iterable of `DimensionElement` or `str`.
+            The other set from which to draw potential result elements.
+
+        Returns
+        -------
+        result : `DimensionNameSet` or `DimensionSet`
+            A new set containing elements ``self`` or ``other``, but not both.
+            A full `DimensionSet` is returned if any argument is a full
+            `DimensionSet` or `DimensionGraph`.
+        """
+        names = self.names ^ toNameSet(other).names
+        universe = getattr(other, "universe", None)
+        if universe is not None:
+            return DimensionSet(universe, names)
+        else:
+            return DimensionNameSet(names)
+
+    def difference(self, other):
+        """Return a new set containing all elements that are in ``self``
+        but not other.
+
+        Parameters
+        ----------
+        other : iterable of `DimensionElement` or `str`.
+            The other set containing elements that should not be included
+            in the result.
+
+        Returns
+        -------
+        result : `DimensionNameSet` or `DimensionSet`
+            A new set containing elements in ``self`` but not ``other``.
+            A full `DimensionSet` is returned if any argument is a full
+            `DimensionSet` or `DimensionGraph`.
+        """
+        names = self.names - toNameSet(other).names
+        universe = getattr(other, "universe", None)
+        if universe is not None:
+            return DimensionSet(universe, names)
+        else:
+            return DimensionNameSet(names)
+
+    # Operators that return sets are only enabled for DimensionNameSet
+    # operands on both sides, to avoid confusion about return types.
+
+    def __or__(self, other):
+        if isinstance(other, DimensionNameSet):
+            return self.union(other)
+        return NotImplemented
+
+    def __and__(self, other):
+        if isinstance(other, DimensionNameSet):
+            return self.intersection(other)
+        return NotImplemented
+
+    def __xor__(self, other):
+        if isinstance(other, DimensionNameSet):
+            return self.symmetric_difference(other)
+        return NotImplemented
+
+    def __sub__(self, other):
+        if isinstance(other, DimensionNameSet):
+            return self.difference(other)
+        return NotImplemented
```

Comment:
I think there would need to be a third (abstract) base class for both, because they don't actually share any state attributes.  I'll at least take a look at that and see if it improves things; overall I'm not *too* concerned about duplication here because it's mostly duplication of trivial things, not tricky logic.